### PR TITLE
(Online shop) sale-items, order Checkout & order status, 

### PR DESCRIPTION
**Pull request by adisa39**
- Adapted models to the updated schema document for Order, OrderItem & Payment.
- Revamp order creation and order update logics to handle order complete.
- Added API endpoints to handle order and other items status change
- Revamp order checkout logics to create and update payment record on order checkout.
- Created PaymentCrossreference model as per updated schema document.

### CODE REVIEW

**Comment by DarinHajou** [comment]
I ran some tests locally and noticed the deleteOrder tests were failing due to a couple mismatches. I made these two changes and all tests passed. 

Change this in specs:

  params: { order_id: '26f5a0f2a86d1f9f3b7e4e81' }

To this:

 params: { id: '26f5a0f2a86d1f9f3b7e4e81' }
 
And change this in the controller:
 
 res.status(200).json({ message: "Order deleted successfully" });

To this:

res.status(200).json({ 
  message: "Order deleted successfully", 
  deletedOrder 
});

**Review by DarinHajou — APPROVED** [review]
Great work on this Adisa, really impressive progress. Just one small, picky thing: in paymentController.ts, the logic for updating the Payment and Order documents (after approval, completion, or handling incomplete payments) is repeated in a few places. Might be worth extracting that into a helper or service method

**Comment by adisa39** [comment]
> Great work on this Adisa, really impressive progress. Just one small, picky thing: in paymentController.ts, the logic for updating the Payment and Order documents (after approval, completion, or handling incomplete payments) is repeated in a few places. Might be worth extracting that into a helper or service method

Thanks @DarinHajou. We can definitely have some codes optimization and cleanup to be done... we can just finish the features testing first. 

For the membership payment, you can just tweak the payment metadata types and toggle payment_type to handle membership subscription, I believe this is a simplified fashion for the two payment used cases.

**Comment by swoocn** [comment]
It seems the **pi-backend** dependency has brought in some additional vulnerabilities.. ugh! 😑

image

File: src/controllers/orderController.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    res.status(200).json(order);
+  } catch (error) {
+    logger.error("Error fetching order: ", error);
+    res.status(500).json({ message: "Failed to fetch order" });
+  }
+};
+
+export const createOrder = async (req: Request, res: Response) => {
+  try {
+    const { orderId, orderData, orderItems } = req.body;
```
Comment (swoocn) [inline src/controllers/orderController.ts:47]: I don't believe `orderId` is used here. 🤔

File: src/controllers/orderController.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      return res.status(404).json({ message: "No items found for this order" });
+    }
+
+    res.status(200).json(result);
+  } catch (error) {
+    res.status(500).json({ message: "Failed to fetch order items" });
+  }
+};
+
+export const updateCompleteOrder = async (req: Request, res: Response) => {
```
Comment (swoocn) [inline src/controllers/orderController.ts:84]: Hi @adisa39 - what are your thoughts about amending this function to support other status transitions other than just completed?  Perhaps rename it to **updateOrderStatus** similar to `updateOrderItemStatus` function..? 🤔

**Comment by swoocn** [comment]
Please ensure filename casing is consistent across packages. 
For example, files in models/enum should follow camelCase, whereas files in models/ should use standard case.  Thanks @adisa39!

File: src/controllers/paymentController.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
+import { Request, Response } from "express";
+import axios from "axios";
+import { platformAPIClient } from "../config/platformAPIclient";
+import { cancelOrder, updatePaidOrder } from "../services/order.service";
+import { IUser, PaymentDataType } from "../types";
+import logger from "../config/loggingConfig";
+import { cancelPayment, completePayment, createA2UPayment, createOrUpdatePaymentCrossReference, createPayment, getPayment } from "../services/payment.service";
```
Comment (swoocn) [inline src/controllers/paymentController.ts:7]: I don't think `createPayment` import is being used in this file.

File: src/controllers/paymentController.ts [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+import { PaymentType } from "../models/enums/paymentType";
+import { U2UPaymentStatus } from "../models/enums/u2uPaymentStatus";
+import OrderCheckout from "../helpers/orderCheckout";
+
+interface PaymentInfo {
```
Comment (swoocn) [inline src/controllers/paymentController.ts:12]: This same interface already exists in the **types** file.

File: src/controllers/paymentController.ts [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+  identifier: string;
+  transaction?: {
+    txid: string;
+    _link: string;
+  };
+}
+
+export const onIncompletePaymentFound = async ( req:Request, res:Response ) => {
+  const currentUser = req.currentUser;
```
Comment (swoocn) [inline src/controllers/paymentController.ts:21]: `currentUser `is not being used in this controller function.

File: src/controllers/paymentController.ts [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    logger.info("incomplete payment data: ", payment);
+
+    const incompletePayment = await getPayment(paymentId);
+
+    if (!incompletePayment) {
+      logger.warn("no incomplete payment")
+      return res.status(400).json({ message: "finding incomplement payment failed" });
+    }
+
+    const horizonResponse = await axios.create({ timeout: 20000 }).get(txURL!);
```
Comment (swoocn) [inline src/controllers/paymentController.ts:37]: Stellar Horizon API 🤔

File: src/controllers/paymentController.ts [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+    await platformAPIClient.post(`/v2/payments/${paymentId}/complete`, {
+      txid,
+    });
+
+    return res.status(200).json({
+      message: `Completed payment from incomplete payment with id ${paymentId}`,
+    });
+  } catch (error: any) {
+    logger.info("error while finding incomplete payment: ", error.message);
```
Comment (swoocn) [inline src/controllers/paymentController.ts:60]: It's strongly recommended to log the full error object, not just the message, so we can capture the complete stack trace for more effective backend debugging.

File: src/controllers/paymentController.ts [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+    return res
+      .status(500)
+      .json(
+        { success: false, error: error.message || error, message: "error completing pi payment" }
```
Comment (swoocn) [inline src/controllers/paymentController.ts:65]: It's safer to log the full error internally but only return a generic message to the client.  This helps avoid leaking sensitive information given the error.

File: src/controllers/paymentController.ts [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+    await platformAPIClient.get(
+      `/v2/payments/${paymentId}`
+    );
+
+    // complete payment status to paid on sucessfull payment
+    const completedPayment = await completePayment(paymentId, txid );
+    logger.info("payment record completed")
+
+    if (completedPayment && completedPayment.payment_type===PaymentType.BuyerCheckout) {
```
Comment (swoocn) [inline src/controllers/paymentController.ts:136]: We can refactor this conditional to `if (completePayment?.payment_type === PaymentType.BuyerCheckout) { }`

File: src/controllers/paymentController.ts [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    return res.status(200).json({
+      status: "ok",
+      success: true,
+      message: `Approved payment with id ${paymentId}`,
+      // order: newOrder, // Optional: Return the created/updated order
+    });
+
+  } catch (error: any) {
+    return res.status(500).json({
+      status: "not ok",
```
Comment (swoocn) [inline src/controllers/paymentController.ts:114]: These status attributes are redundant and can be removed.

File: src/controllers/paymentController.ts [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    await platformAPIClient.post( `/v2/payments/${paymentId}/cancel` );
+    logger.info( "response from Pi server on payment cancellation " );
+
+    return res
+      .status(200)
+      .json({ success: true, message: `Cancelled the payment ${paymentId}` });
+      
+  } catch (error: any) {
+    logger.error("Error while canceling transaction: ", error.message);
+    return res.status(400).json({ success: false, message: "Internal Server Error" });
```
Comment (swoocn) [inline src/controllers/paymentController.ts:208]: ~400~ -> 500

File: src/controllers/paymentController.ts
Comment (swoocn) [inline src/controllers/paymentController.ts:1]: TODO:  Apply separation of concerns and extract business logic into service layer; controller layers should primarily handle request/ response logic while business logic and workflows should be in the service layer.

File: src/services/payment.service.ts [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    const payment = new Payment({
+      pi_payment_id: paymentData.piPaymentId,
+      user_id: paymentData.userId,
+      amount: paymentData.amount,
+      paid: false,
+      memo: paymentData.memo,
+      payment_type: paymentData.paymentType,
+      cancelled: false,
+    });
+    const newPayment = await payment.save();
```
Comment (swoocn) [inline src/services/payment.service.ts:30]: `.save()` will already throw an error if it fails; so the following conditional is redundant and unnecessary.

File: src/controllers/paymentController.ts (hunk 6 above)
Comment (adisa39) [reply src/controllers/paymentController.ts:34]: any observation on this? the code snippet is from PiSDK

File: src/controllers/paymentController.ts (hunk 7 above)
Comment (adisa39) [reply src/controllers/paymentController.ts]: the full error log is very long but I will modify as indicated

File: src/controllers/paymentController.ts (hunk 10 above)
Comment (adisa39) [reply src/controllers/paymentController.ts]: It is from PiSDK but I will change it and see if it doesn't break anything from the FE callback functions.

File: src/controllers/paymentController.ts
Comment (adisa39) [reply src/controllers/paymentController.ts:1]: That is why I suggested robust testing of the payment, it will be easier to refactor to cover edge cases and errors at a go. Some edge cases may actually require complete restructuring of the codes. This will do for now.

File: src/controllers/paymentController.ts (hunk 6 above)
Comment (swoocn) [reply src/controllers/paymentController.ts:34]: Just an observation; you're good. 👌

File: src/controllers/paymentController.ts (hunk 7 above)
Comment (swoocn) [reply src/controllers/paymentController.ts]: IMO, more detailed error logging is always beneficial, @adisa39.  
I'm making ongoing PR adjustments on my end, so no need for you to update anything for now.  Please hold off until the final PR amendments are in.  I will let you know.  Thanks!

File: src/controllers/paymentController.ts
Comment (swoocn) [reply src/controllers/paymentController.ts:1]: I am working on making the adjustments as it will be easier to write unit tests for the various edge cases so that robust testing will be more favorable as per my comment @ [https://app.clickup.com/t/86c12kbuy?comment=90150120973484&threadedComment=90150120998559](https://app.clickup.com/t/86c12kbuy?comment=90150120973484&threadedComment=90150120998559).

**Comment by swoocn** [comment]
I'm making ongoing PR adjustments on my end, so no need for you to update anything for now.  Please hold off until the final PR amendments are in.  I will let you know.  Thanks @adisa39!

File: src/services/order.service.ts [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+        subtotal: item.quantity * parseFloat(sellerItem.price.toString()),
+        status: OrderItemStatusType.Pending,
+      };
+    });
+
+    // Insert all order items in bulk
+    await OrderItem.insertMany(bulkOrderItems, { session });
+
+    await session.commitTransaction();
+    session.endSession();
```
Comment (swoocn) [inline src/services/order.service.ts:70]: Perhaps we should consider adding a finally block to ensure `session.endSession()` is always called, regardless of the outcome.

File: src/services/order.service.ts [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+    const updatedOrder = await Order.findByIdAndUpdate(
+      orderId, 
+      { status: orderStatus },
+      { new: true }
+    ).exec();
+    
+    if (!updatedOrder) return null;
+
+    // Return order items
+    const orderItems = await getOrderItems(orderId);
```
Comment (swoocn) [inline src/services/order.service.ts:208]: TODO: Refactor opportunity to remove coupling; perhaps we could let the caller decide in the controller layer if fetching full order details is necessary..?

File: src/controllers/paymentController.ts
Comment (swoocn) [reply src/controllers/paymentController.ts:1]: WIP ⚡

File: src/services/order.service.ts [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+      .lean();
+    return orders;
+
+  } catch (error: any) {
+    logger.error(`Failed to get seller orders for Pi UID ${ piUid }: ${ error.message }`);
+    throw error;
+  }
+};
+
+export const getBuyerOrdersById = async (piUid: string) => {
```
Comment (swoocn) [inline src/services/order.service.ts:134]: This function is essentially a copy of the previous `getSellerOrdersById()`.  Candidate for consolidation into one function?

**Review by amt-fredrick-amoako — APPROVED** [review]

**Comment by swoocn** [comment]
Hi @adisa39 – It looks like your recent payment-related change caused a unit test to fail, which broke the build.
No worries—I went ahead and fixed the test for you.
